openshift/openshift-ansible-contrib
# [WIP] Support external/pre-provisioned authoritative cluster DNS

#### What does this PR do?

* Allow sending updates to multiple (based on the `full_dns_domain` fact) private and public views via 
  mixed external/in-stack DNS servers (i.e.: provisioned for internal use and external for external use).
* Support dynamic updates to multiple clusters views served by multiple DNS servers (see the 
  `external_nsupdate_keys` dict for details). Auto-evaluate used nsupdate keys for the instack DNS 
  server, when using a mixed instack/external DNS servers.
* Also allow to not provision in-stack DNS servers at all and just go with fully external DNS servers.

#### How should this be manually tested?

Fully external DNS server:

* Deploy s DNS server (f.e. by [this guide](https://github.com/openshift/openshift-ansible-contrib/tree/master/reference-architecture/osp-dns)) and populate an inventory file (e.g. `inventory/external_dns.yaml`) with its nsupdate data
* Run provision openstack with app nodes count of 1 and dns nodes count of 0, and add 
  `-e@inventory/external_dns.yaml` for the ansible-playbook commands used
* Finish e2e (it shall pass)
* Run provision openstack with app nodes count of 2 and dns nodes count of 0 and add 
  `-e@inventory/external_dns.yaml` for the ansible-playbook commands used
* Finish e2e (it shall pass)
* Check if all app nodes' FQDNs are resolvable via your external DNS

Mixed instack/external DNS:

* Deploy s DNS server (f.e. by [this guide](https://github.com/openshift/openshift-ansible-contrib/tree/master/reference-architecture/osp-dns)) and populate an inventory file (e.g. `inventory/external_dns.yaml`) with its nsupdate data as a public view, for a private view add this:
```
...
private-openshift.example.com:
  instack: true
  public_access: false # use true if a control node does not belong to the admin (private) cluster network
```
* Run provision openstack with app nodes count of 1 and dns nodes count of 1, and add
  `-e@inventory/external_dns.yaml` for the ansible-playbook commands used
* Finish e2e (it shall pass)
* Run provision openstack with app nodes count of 2 and dns nodes count of 1 and add
  `-e@inventory/external_dns.yaml` for the ansible-playbook commands used
* Finish e2e (it shall pass)
* Check if all app nodes' public FQDNs are resolvable via your external DNS and the private DNS server 
  is used for a private view.

#### Is there a relevant Issue open for this?

#### Who would you like to review this?
cc: @tomassedovic @oybed PTAL
